src/ORM/ArrayList.php

```diff
@@ -731,7 +742,21 @@ protected function filterOrExclude(array $filters, bool $inclusive = true, bool
             foreach ($filters as $filterKey => $filterValue) {
                 /** @var SearchFilter $searchFilter */
                 $searchFilter = $searchFilters[$filterKey];
-                $hasMatch = $searchFilter->matches($this->extractValue($item, $searchFilter->getFullName()) ?? '');
+                $extractedValue = $this->extractValue($item, $searchFilter->getFullName());
+                $hasMatch = $searchFilter->matches($extractedValue);
+
+                // If we didn't get a match, check if we need to do a legacy null
+                // comparison for backwards compatibility
+                if (!$hasMatch
+                    && ($searchFilter instanceof ExactMatchFilter)
+                    && ($hasNullFilter || $extractedValue === null)
+                ) {
+                    $hasMatch = $this->performLegacyNullMatch($extractedValue, $filterValue);
+                    if (in_array('not', $searchFilter->getModifiers())) {
+                        $hasMatch = !$hasMatch;
+                    }
+                }
+
```

Comment:
Instead of just casting null to empty string, we now first let the search filter do its thing. Only if it _doesn't_ get a match, do we then check if we need to do a legacy null comparison.

Reply:
Changed this to doing the null check first, due to some edge cases. But if the object value isn't null and we don't get a match, we still pass it through to the search filter afterward.